src/notifications/mod.rs

```diff
@@ -0,0 +1,149 @@
+mod discord;
+
+use crate::notifications::discord::{build_discord_payload, is_discord_webhook};
+use chrono::prelude::*;
+use inflections::case::{to_constant_case, to_title_case};
+use log::{debug, error, info};
+use reqwest::blocking::RequestBuilder;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use std::env::VarError;
+use std::{env, fmt};
+
+pub const WEBHOOK_URL: &str = "WEBHOOK_URL";
+
+#[derive(Debug, Copy, Clone, Deserialize, Serialize)]
+pub enum NotificationEventType {
+  Broadcast,
+  //
+  Updating,
+  UpdateSuccessful,
+  UpdateFailed,
+  //
+  Starting,
+  StartSuccessful,
+  StartFailed,
+  //
+  Stopping,
+  StopSuccessful,
+  StopFailed,
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct NotificationMessage {
+  event_type: NotificationEventType,
+  event_message: String,
+  timestamp: String,
+}
+
+impl fmt::Display for NotificationEventType {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "{:?}", self)
+  }
+}
+
+fn fetch_webhook_url() -> Result<String, VarError> {
+  env::var(WEBHOOK_URL)
+}
+
+fn is_webhook_enabled() -> bool {
+  fetch_webhook_url().is_ok()
+}
+
+pub trait NotificationEvent {
+  fn handle_request(&self, request: RequestBuilder);
+  fn create_notification(&self) -> NotificationMessage;
+  fn build_request(&self, webhook_url: &str) -> RequestBuilder;
+  fn send_custom_notification(&self, webhook_url: &str, message: &str);
+  fn send_notification(&self);
+}
+
+impl NotificationEvent for NotificationEventType {
+  fn handle_request(&self, request: RequestBuilder) {
+    let response = request.send();
+    if let Ok(parsed_response) = response {
+      let response_status = parsed_response.status();
+      let response_message = parsed_response.text().unwrap();
+      match response_status.as_u16() {
+        204 | 201 => info!("[{}]: Webhook message sent successfully!", self.to_string()),
+        _ => error!("Request failed! {}, {}", response_status, response_message),
+      }
+    } else {
+      error!(
+        "[{}]: Error with webhook! Status {}",
+        self.to_string(),
+        response
+          .err()
+          .unwrap()
+          .status()
+          .unwrap_or(StatusCode::INTERNAL_SERVER_ERROR)
+          .as_str()
+      );
+    }
+  }
+  fn create_notification(&self) -> NotificationMessage {
+    let name = self.to_string();
+    NotificationMessage {
+      event_type: self.to_owned(),
+      event_message: format!("Server Status: {}", to_title_case(name.as_str())),
+      timestamp: Local::now().format("%m/%d/%Y %H:%M:%S %Z").to_string(),
+    }
+  }
```

Comment:
Im stuck ;-; I have all the other portions done but this one is giving me greif.  When I have it as below I cannot pass self into it..

```rs
impl From<NotificationEvent> for NotificationMessage {
  fn from(event: NotificationEvent) -> Self {
    NotificationMessage {
      event_type: event,
      event_message: format!(
        "Server Status: {}",
        to_title_case(event.to_string().as_str())
      ),
      timestamp: Local::now().to_rfc3339(),
    }
  }
}
```

Like so 
```rs
 let mut notification = NotificationMessage::from(self);
    notification.event_message = message.to_string();
    debug!("Webhook enabled, sending notification {}", self);

```

But when I change it to a barrowed variable

```rs
impl From<&NotificationEvent> for NotificationMessage {
  fn from(event: &NotificationEvent) -> Self {
    NotificationMessage {
      event_type: event,
      event_message: format!(
        "Server Status: {}",
        to_title_case(event.to_string().as_str())
      ),
      timestamp: Local::now().to_rfc3339(),
    }
  }
}
```

Self works just fine being passed in but `event_type: event` does not work because it is a barrowed event...

Any solutions I could pursue?

Reply:
If you want you can just ignore this one for now and push it how it was before and then I can take a look at it later. I'm still planning on doing that pass to clean stuff up so I can include trying to get this change working with it. It may take some tweaking on borrowing vs passing in an owned value to get it to work